src/joystick/windows/SDL_dinputjoystick.c

```diff
@@ -555,22 +543,33 @@ typedef struct
 } EnumJoystickPresentData;
 
 static BOOL CALLBACK
-EnumJoystickPresentCallback(const DIDEVICEINSTANCE * pdidInstance, VOID * pContext)
+EnumJoystickPresentCallback(LPCDIDEVICEINSTANCE pDeviceInstance, LPVOID pContext)
 {
+#define CHECK(exp) { if(!(exp)) goto err; }
     EnumJoystickPresentData *data = (EnumJoystickPresentData *)pContext;
     Uint16 vendor = 0;
     Uint16 product = 0;
-    Uint16 version = 0;
+    LPDIRECTINPUTDEVICE8 device = NULL;
+    BOOL ret = DIENUM_CONTINUE;
 
-    if (SDL_memcmp(&pdidInstance->guidProduct.Data4[2], "PIDVID", 6) == 0) {
-        vendor = (Uint16)LOWORD(pdidInstance->guidProduct.Data1);
-        product = (Uint16)HIWORD(pdidInstance->guidProduct.Data1);
-        if (data->vendor == vendor && data->product == product && data->version == version) {
-            data->present = SDL_TRUE;
-            return DIENUM_STOP;
-        }
+    /* We are only supporting HID devices. */
+    CHECK((pDeviceInstance->dwDevType & DIDEVTYPE_HID) != 0);
+
+    CHECK(SUCCEEDED(IDirectInput8_CreateDevice(dinput, &pDeviceInstance->guidInstance, &device, NULL)));
```

Comment:
Creating a device inside of the enumeration is expensive. Have you done performance testing of this with 4+ game controllers connected?

Reply:
I'll do some perfomance checks.

I wonder if I can replace this `IDirectInput8_EnumDevices` in `SDL_DINPUT_JoystickPresent` with `SYS_Joystick` loop. It already have created `DIDEVICEINSTANCE dxdevice` in for DInput devices

Reply:
In general, creating the device instance with every enumeration seems expensive. I don't think we're polling anymore, so maybe this only happens at startup or on device change, which might be okay. Anywhere we can short circuit expensive logic would be a good thing though.

Reply:
done